File: src/main/java/io/github/ozkanpakdil/opentelemetry/ui/TelemetryTableModel.java
package io.github.ozkanpakdil.opentelemetry.ui;

import io.github.ozkanpakdil.opentelemetry.Telemetry;
import io.github.ozkanpakdil.opentelemetry.TelemetryType;
import io.github.ozkanpakdil.opentelemetry.utils.TimeSpan;
import org.jetbrains.annotations.Nullable;

import javax.swing.table.AbstractTableModel;
import java.util.*;

public class TelemetryTableModel extends AbstractTableModel {
    private final String[] columnNames = new String[]{
            "timestamp", "duration", "data"
    };
    private final Class<?>[] columnClass = new Class[]{
            Date.class, TimeSpan.class, Telemetry.class
    };
    private final List<Telemetry> telemetries;

    public TelemetryTableModel() {
        this.telemetries = new ArrayList<>();
    }

    @Override
    public String getColumnName(int column) {
        return columnNames[column];
    }

    @Override
    public Class<?> getColumnClass(int columnIndex) {
        return columnClass[columnIndex];
    }

    @Override
    public int getRowCount() {
        return telemetries.size();
    }

    @Override
    public int getColumnCount() {
        return columnNames.length;
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        Telemetry telemetry = telemetries.get(rowIndex);
        return switch (columnIndex) {
            case 0 -> telemetry.getTimestamp();
            case 1 -> telemetry.getDuration();
            case 2 -> telemetry;
            default -> null;
        };
    }

    public void addRow(Telemetry telemetry) {
        this.telemetries.add(telemetry);
        this.fireTableRowsInserted(this.telemetries.size() - 1, this.telemetries.size() - 1);
    }

    public void addRow(int i, Telemetry telemetry) {
        this.telemetries.add(i, telemetry);
        this.fireTableRowsInserted(this.telemetries.size() - 1, this.telemetries.size() - 1);
    }

    public void setRows(List<Telemetry> telemetries) {
        int previousSize = this.telemetries.size();
        this.telemetries.clear();
        this.fireTableRowsDeleted(0, Math.max(0, previousSize - 1));
        this.telemetries.addAll(telemetries);
        this.fireTableRowsInserted(0, Math.max(0, this.telemetries.size() - 1));
    }

    @Nullable
    public Telemetry getRow(int selectedRow) {
        if (selectedRow < 0)
            return null;
        if (selectedRow >= this.getRowCount())
            return null;
        return telemetries.get(selectedRow);
    }
}

File: src/main/java/io/jeremymorren/opentelemetry/ui/TelemetryTableModel.kt
package io.jeremymorren.opentelemetry.ui

import io.jeremymorren.opentelemetry.Telemetry
import io.jeremymorren.opentelemetry.TelemetryItem
import io.jeremymorren.opentelemetry.TelemetryType
import io.jeremymorren.opentelemetry.utils.TimeSpan
import java.time.Instant
import java.util.*
import javax.swing.table.AbstractTableModel
import kotlin.math.max

class TelemetryTableModel : AbstractTableModel() {
    private val columnNames = arrayOf(
        "timestamp", "duration", "type", "detail"
    )
    private val columnClass = arrayOf<Class<*>>(
        Instant::class.java, TimeSpan::class.java, TelemetryType::class.java, Telemetry::class.java
    )

    private val telemetries: MutableList<TelemetryItem> = ArrayList()

    override fun getColumnName(column: Int): String = columnNames[column]

    override fun getColumnClass(columnIndex: Int): Class<*> = columnClass[columnIndex]

    override fun getRowCount(): Int = telemetries.size

    override fun getColumnCount(): Int = columnNames.size

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any? {
        val telemetry = telemetries[rowIndex]
        return when (columnIndex) {
            0 -> telemetry.timestamp
            1 -> telemetry.duration
            2 -> telemetry.type
            3 -> telemetry.telemetry
            else -> null
        }
    }

    fun addRow(telemetry: TelemetryItem) {
        telemetries.add(telemetry)
        this.fireTableRowsInserted(telemetries.size - 1, telemetries.size - 1)
    }

    fun addRow(i: Int, telemetry: TelemetryItem) {
        telemetries.add(i, telemetry)
        this.fireTableRowsInserted(telemetries.size - 1, telemetries.size - 1)
    }

    fun setRows(telemetries: List<TelemetryItem>) {
        val previousSize = this.telemetries.size
        this.telemetries.clear()
        this.fireTableRowsDeleted(0, max(0.0, (previousSize - 1).toDouble()).toInt())
        this.telemetries.addAll(telemetries)
        this.fireTableRowsInserted(0, max(0.0, (this.telemetries.size - 1).toDouble()).toInt())
    }

    fun getRow(selectedRow: Int): TelemetryItem? {
        if (selectedRow < 0) return null
        if (selectedRow >= this.rowCount) return null
        return telemetries[selectedRow]
    }
}
